draco/analysis/powerspec.py

```diff
@@ -0,0 +1,1504 @@
+        wl = units.c / (nu_c * 1e6)  # m
+
+        # find out the kperp bins
+        u_min_lambda = self.bl_min / wl
+        u_max_lambda = self.bl_max / wl
+        kperp_min = u_to_kperp(u_min_lambda, redshift, ps.cosmology)
+        kperp_max = u_to_kperp(u_max_lambda, redshift, ps.cosmology)
+
+        if self.logbins_2D:
+            kperp = np.logspace(np.log10(kperp_min), np.log10(kperp_max), self.Nbins_2D)
+        else:
+            kperp = np.linspace(kperp_min, kperp_max, self.Nbins_2D)
+
+        # Take the center of the kperp bins
+        kperp_cent = 0.5 * (kperp[1:] + kperp[:-1])
+
+        # Dereference the required datasets
+        ps_3D = ps.ps3D[:].local_array
+
+        if self.weight is None:
+            weight = np.ones_like(ps_3D)
+        else:
+            # input weight is variance and we are taking the
+            # inverse to have a inverse variance weight
+            self.weight.redistribute("delay")
+            weight = tools.invert_no_zero(self.weight.ps3D[:].local_array)
+
+        # Define the 2D power spectrum container
+        pspec_2D = containers.PowerSpectrum2D(
+            pol=pol,
+            delay=delay,
+            kperp=kperp_cent,
+            attrs_from=ps,
+            cosmology=ps.cosmology,
+        )
+        pspec_2D.redistribute("delay")
+        pspec_2D.ps2D[:] = 0.0
+        pspec_2D.k_parallel[:] = kpar
+        pspec_2D.kperp[:] = kperp_cent
+
+        # save the delay cut value in attrs
+        pspec_2D.attrs["delay_cut"] = self.delay_cut
+
+        # loop over polarization
+        for pp, psr in enumerate(pol):
+            self.log.debug(f"Estimating 2D power spectrum for pol: {psr}")
+
+            # loop over k_parallel axis
+            for lde, de in pspec_2D.ps2D[:].enumerate(axis=1):
+                slc = (pp, lde, slice(None), slice(None))
+                data = np.ascontiguousarray(ps_3D[slc])
+                W = np.ascontiguousarray(weight[slc])
+
+                # keep the non-zero visibilities between minimum and maximum baseline
+                # for each delay channel and take the flatten array; i.e, the output is (nvis),
+                # where nvis is number of vis between two limit
+
+                ps3D_flat, uu, vv = reshape_data_cube(
+                    data, u, v, u_min_lambda, u_max_lambda
+                )
+
+                # do the same for uv-spatial mask and weight
+                # and apply the mask before passing to 2D ps estimate
+                mask_flat, _, _ = reshape_data_cube(
+                    uv_mask, u, v, u_min_lambda, u_max_lambda
+                )
+
+                weight_flat, _, _ = reshape_data_cube(
+                    W, u, v, u_min_lambda, u_max_lambda
+                )
+
+                # Cylindrical 2D power spectrum
+                pspec_2D.ps2D[pp, de], pspec_2D.weight[pp, de] = get_2d_ps(
+                    ps3D_flat[mask_flat],
+                    weight=weight_flat[mask_flat]
```

Comment:
You say you are squaring here to get inverse variance, but the docstring in the `setup` method seems to suggest the quantity being saved to the `weight` container is already a variance.  Is the quantity in the weight container actually a standard deviation?  If so, then the docstring should be updated.  If not, then I think this `**2` needs to be removed.